README.md
[![Build Status](https://travis-ci.org/sertansenturk/ahenkidentifier.svg?branch=master)](https://travis-ci.org/sertansenturk/ahenkidentifier) [![codecov.io](https://codecov.io/github/sertansenturk/ahenkidentifier/coverage.svg?branch=master)](https://codecov.io/github/sertansenturk/ahenkidentifier?branch=master) [![Code Climate](https://codeclimate.com/github/sertansenturk/ahenkidentifier/badges/gpa.svg)](https://codeclimate.com/github/sertansenturk/ahenkidentifier) [![GitHub version](https://badge.fury.io/gh/sertansenturk%2Fahenkidentifier.svg)](https://badge.fury.io/gh/sertansenturk%2Fahenkidentifier) [![DOI](https://zenodo.org/badge/21104/sertansenturk/ahenkidentifier.svg)](https://zenodo.org/badge/latestdoi/21104/sertansenturk/ahenkidentifier) [![License: AGPL v3](https://img.shields.io/badge/License-AGPL%20v3-ff69b4.svg)](http://www.gnu.org/licenses/agpl-3.0)

# ahenkidentifier
Identifies the ahenk (transposition) of a makam music recording given the tonic frequency and the symbol (or the makam)

If you are using ahenkidentifier in your work, please cite:

> Şentürk, S. (2016). [Computational Analysis of Audio Recordings and Music Scores for the Description and Discovery of Ottoman-Turkish Makam Music](compmusic.upf.edu/senturk2016thesis). PhD thesis, Universitat Pompeu Fabra, Barcelona, Spain.

Usage
-------

```python
from ahenkidentifier.ahenkidentifier import AhenkIdentifier

ahenk = AhenkIdentifier.identify(tonic_freq, makam)
# or 
ahenk = AhenkIdentifier.identify(tonic_freq, tonic_symbol)
```

The inputs are:
```python
# tonic_freq 	    :	The frequency of the tonic in Hz.
# makam/tonic_symbol:	The algorithm can either accept the makam-slug or 
#					    the note symbol of the tonic in SymbTr format as a string (e.g. B4b1).
```
For the makam-slug names, check the json file in the data folder. The slugs are the same with the ones in the filenames of the scores in the [SymbTr](https://github.com/MTG/SymbTr) collection. The tonic symbols are notated as \[Note pitch-class\]\[Octave\](Accidental Symbol(Holderian Comma)), e.g. B4b1

The output is:
```python
# ahenk 		  :	The name of the ahenk as a unicode string and a dictionary of computed stats
```

Please refer to [demo.ipynb](https://github.com/sertansenturk/ahenkidentifier/blob/master/demo.ipynb) for an interactive demo.

Installation
-------

If you want to install ahenkidentifier, it is recommended to install the package and dependencies into a virtualenv. In the terminal, do the following:

    virtualenv env
    source env/bin/activate
    python setup.py install

If you want to be able to edit files and have the changes be reflected, then install the repository like this instead:

    pip install -e .

Issues
-------
The "Bolahenk Nısfiye" ahenk, which is an octave higher than the default ahenk ("Bolahenk"), is omitted. In solo performances, automatic identification of this ahenk requires the frequency range of the instrument being performed. Moreover, it is ambiguous to distinguish it from "Bolahenk" in multi-instrument recordings.

Authors
-------
Sertan Şentürk
contact@sertansenturk.com


Changelog.md
#### ahenkidentifier v1.5.2
 - Input tonic frequency is validation added.
 - The tonic symbol of is returned as "unknown", if the tonic symbol of a makam is not given in the file [tonic.json](https://github.com/sertansenturk/ahenkidentifier/commit/d617f74e6b358fa0f42f81d44f049f3a219a4e86)
 
#### ahenkidentifier v1.5.1
 - Corrected Sultaniyegah tonic symbol from D5 to D4

#### ahenkidentifier v1.5.0
 - Changed the makam/tonic symbol error from IOError to ValueError

#### ahenkidentifier v1.4.0
 - Added more unittests
 - Code coverage integration
 - Code climate integration

#### ahenkidentifier v1.3.0
 - Merged the two outputs into a single dictionary
 - Added missing MANIFEST file and the data packages when installing from github

#### ahenkidentifier v1.2.0
 - Refactored the methods into the AhenkIdentifier object

#### ahenkidentifier v1.1.1
 - Fixed input tonic_symbol/makam_slug checking

#### ahenkidentifier v1.1
 - Python 3.3, 3.4, 3.5 compatibility

#### ahenkidentifier v1.0.1
 - Fixes the ahenk identification bug due to erroneous cent distance computation (Issue: [MTG/dunya#307](https://github.com/MTG/dunya/issues/307))

#### ahenkidentifier v1.0
 - First public release


unittests/ahenk_test.py
from ahenkidentifier.AhenkIdentifier import AhenkIdentifier
import numpy as np
import json


def test_ahenk():
    success = True

    # load the correct results
    correct_ahenks = json.load(open('unittests/correct_ahenks.json'))

    # parameters
    start_freq = 16  # C0 note
    num_octaves = 10
    cent_step = 10.0  # cents
    dummy_note_symbol = 'A4'

    # get the frequencies to get ahenk
    cc = np.arange(0, num_octaves, cent_step / AhenkIdentifier.CENTS_IN_OCTAVE)
    freqs = [int(i) for i in start_freq * np.power(2.0, cc)]

    # compute
    for freq in freqs:
        ahenk = AhenkIdentifier.identify(freq, dummy_note_symbol)
        if not ahenk['name'] == correct_ahenks[str(freq)][0]:
            print("Mismatch in " + dummy_note_symbol + ' = ' + str(freq))
            success = False

    assert success


ahenkidentifier/unittests/ahenk_test.py
from ahenkidentifier.ahenkidentifier import AhenkIdentifier
import numpy as np
import json
import os

_curr_folder = os.path.dirname(os.path.abspath(__file__))


def test_ahenk_with_tonic_symbol_input():
    dummy_tonic_symbol = 'A4'
    assert t_wrapper(dummy_tonic_symbol)


def test_ahenk_with_makam_input():
    dummy_makam = 'huseyni'
    assert t_wrapper(dummy_makam)


def test_ahenk_with_makam_with_unknown_tonic_input():
    dummy_makam = 'dusems'
    wrong_str_err = "The tonic of this makam is not known."
    correct_err_msg = t_error_wrapper(dummy_makam, KeyError, wrong_str_err)

    assert correct_err_msg


def test_ahenk_with_random_str():
    dummy_tonic_symbol = 'eheh'
    wrong_str_err = "The second input has to be a tonic symbol or a " \
                    "makam slug!"
    correct_err_msg = t_error_wrapper(dummy_tonic_symbol, ValueError,
                                      wrong_str_err)

    assert correct_err_msg


def t_error_wrapper(wrong_str, err_type, wrong_str_err):
    correct_err_msg = False
    try:
        t_wrapper(wrong_str)
    except err_type as e:
        # str(e) returns the error message with trailing '', that's why we
        # check the validity with "in"
        correct_err_msg = wrong_str_err in str(e)

    return correct_err_msg


def t_wrapper(dummy_str):
    # load the correct results
    ahenk_file = os.path.join(_curr_folder, 'correct_ahenks.json')
    correct_ahenks = json.load(open(ahenk_file, 'r'))

    # parameters
    start_freq = 16.0  # C0 note
    num_octaves = 10
    cent_step = 10.0  # cents

    # get the frequencies to get ahenk
    cc = np.arange(0, num_octaves, cent_step / AhenkIdentifier.CENTS_IN_OCTAVE)
    freqs = [int(i) for i in start_freq * np.power(2.0, cc)
             if 20.0 <= i <= 20000.0]  # restrict to 20-20000Hz for assertion

    # compute
    success = True
    for freq in freqs:
        ahenk = AhenkIdentifier.identify(freq, dummy_str)
        if not ahenk['name'] == correct_ahenks[str(freq)][0]:
            print("Mismatch in " + dummy_str + ' = ' + str(freq))
            success = False

    return success


ahenkidentifier/AhenkIdentifier.py
import os
import json
import numpy as np
from future.utils import iteritems


class AhenkIdentifier(object):
    CENTS_IN_OCTAVE = 1200  # cents

    @classmethod
    def identify(cls, tonic_freq, symbol_in):
        tonic_dict = cls._get_dict('tonic')
        ahenks = cls._get_dict('ahenk')

        # get the tonic symbol and frequency
        tonic_symbol, tonic_bolahenk_freq, makam = cls._get_tonic_symbol(
            symbol_in, tonic_dict)

        # get the transposition in cents, rounded to the closest semitone
        cent_dist = cls._hz_to_cent(tonic_freq, tonic_bolahenk_freq)
        mod_cent_dist = np.mod(cent_dist, cls.CENTS_IN_OCTAVE)

        # if the distance is more than 1150 cents wrap it to minus
        # so it will be mapped to 0 cents
        mod_cent_dist = (mod_cent_dist if mod_cent_dist < 1150
                         else mod_cent_dist - 1200)

        mod_cent_approx = int(np.round(mod_cent_dist * 0.01) * 100)
        mod_cent_dev = mod_cent_approx - mod_cent_dist
        abs_mod_cent_dev = abs(mod_cent_dev)

        # create the stats dictionary
        distance_to_bolahenk = {
            'performed': {'value': mod_cent_dist.tolist()[0], 'unit': 'cent'},
            'theoretical': {'value': mod_cent_approx, 'unit': 'cent'}}
        ahenk_dict = {'name': '', 'slug': '', 'makam': makam,
                      'tonic_symbol': tonic_symbol,
                      'distance_to_bolahenk': distance_to_bolahenk,
                      'deviation': {'value': mod_cent_dev.tolist()[0],
                                    'unit': 'cent'},
                      'abs_deviation': {'value': abs_mod_cent_dev.tolist()[0],
                                        'unit': 'cent'}}

        # get the ahenk
        for ahenk_slug, val in iteritems(ahenks):
            if val['cent_transposition'] == mod_cent_approx:
                ahenk_dict['name'] = val['name']
                ahenk_dict['slug'] = ahenk_slug
                return ahenk_dict

    @staticmethod
    def _get_tonic_symbol(symbol_in, tonic_dict):
        if symbol_in in tonic_dict.keys():  # tonic symbol given
            tonic_symbol = symbol_in
            makam = None
            tonic_bolahenk_freq = tonic_dict[symbol_in]['bolahenk_freq']
        else:  # check if the makam name is given
            makam = symbol_in
            tonic_symbol, tonic_bolahenk_freq = \
                AhenkIdentifier._get_tonic_symbol_from_makam(symbol_in,
                                                             tonic_dict)
            if not tonic_bolahenk_freq:
                raise IOError("The second input has to be a tonic " +
                              "symbol or a makam slug!")

        return tonic_symbol, tonic_bolahenk_freq, makam

    @staticmethod
    def _get_tonic_symbol_from_makam(symbol_in, tonic_dict):
        tonic_bolahenk_freq = tonic_symbol = None
        for sym, val in iteritems(tonic_dict):
            if symbol_in in val['makams']:
                tonic_symbol = sym
                tonic_bolahenk_freq = val['bolahenk_freq']
                if not tonic_bolahenk_freq:  # tonic symbol is not known
                    raise KeyError("The tonic of this makam is not known.")
                break
        return tonic_symbol, tonic_bolahenk_freq

    @staticmethod
    def _get_dict(dict_type):
        dict_file = os.path.join(os.path.dirname(
            os.path.abspath(__file__)), 'data', dict_type + '.json')
        return json.load(open(dict_file, 'r'))

    @staticmethod
    def _hz_to_cent(hz_track, ref_freq):
        """--------------------------------------------------------------------
        Converts an array of Hertz values into cents.
        -----------------------------------------------------------------------
        hz_track : The 1-D array of Hertz values
        ref_freq    : Reference frequency for cent conversion
        --------------------------------------------------------------------"""
        hz_track = np.array(hz_track)

        # The 0 Hz values are removed, not only because they are meaningless,
        # but also logarithm of 0 is problematic.
        return np.log2(hz_track[hz_track > 0] / ref_freq) * 1200.0


ahenkidentifier/ahenkidentifier.py
import os
import json
import numpy as np
from future.utils import iteritems


class AhenkIdentifier(object):
    CENTS_IN_OCTAVE = 1200  # cents

    @classmethod
    def identify(cls, tonic_freq, symbol_in):
        assert 20.0 <= tonic_freq <= 20000.0, "The input tonic frequency " \
                                              "must be between and 20000 Hz"

        tonic_dict = cls._get_dict('tonic')
        ahenks = cls._get_dict('ahenk')

        # get the tonic symbol and frequency
        tonic_symbol, tonic_bolahenk_freq, makam = cls._get_tonic_symbol(
            symbol_in, tonic_dict)

        # get the transposition in cents, rounded to the closest semitone
        cent_dist = cls._hz_to_cent(tonic_freq, tonic_bolahenk_freq)
        mod_cent_dist = np.mod(cent_dist, cls.CENTS_IN_OCTAVE)

        # if the distance is more than 1150 cents wrap it to minus
        # so it will be mapped to 0 cents
        mod_cent_dist = (mod_cent_dist if mod_cent_dist < 1150
                         else mod_cent_dist - 1200)

        mod_cent_approx = int(np.round(mod_cent_dist * 0.01) * 100)
        mod_cent_dev = mod_cent_approx - mod_cent_dist
        abs_mod_cent_dev = abs(mod_cent_dev)

        # create the stats dictionary
        distance_to_bolahenk = {
            'performed': {'value': mod_cent_dist.tolist()[0], 'unit': 'cent'},
            'theoretical': {'value': mod_cent_approx, 'unit': 'cent'}}
        ahenk_dict = {'name': '', 'slug': '', 'makam': makam,
                      'tonic_symbol': tonic_symbol,
                      'distance_to_bolahenk': distance_to_bolahenk,
                      'deviation': {'value': mod_cent_dev.tolist()[0],
                                    'unit': 'cent'},
                      'abs_deviation': {'value': abs_mod_cent_dev.tolist()[0],
                                        'unit': 'cent'}}

        # get the ahenk
        for ahenk_slug, val in iteritems(ahenks):
            if val['cent_transposition'] == mod_cent_approx:
                ahenk_dict['name'] = val['name']
                ahenk_dict['slug'] = ahenk_slug
                return ahenk_dict

    @classmethod
    def _get_tonic_symbol(cls, symbol_in, tonic_dict):
        if symbol_in in tonic_dict.keys():  # tonic symbol given
            tonic_symbol = symbol_in
            makam = None
            tonic_bolahenk_freq = tonic_dict[symbol_in]['bolahenk_freq']
        else:  # check if the makam name is given
            makam = symbol_in
            tonic_symbol, tonic_bolahenk_freq = \
                cls._get_tonic_symbol_from_makam(symbol_in, tonic_dict)
            if not tonic_bolahenk_freq:
                raise ValueError("The second input has to be a tonic symbol "
                                 "or a makam slug!")

        return tonic_symbol, tonic_bolahenk_freq, makam

    @staticmethod
    def _get_tonic_symbol_from_makam(symbol_in, tonic_dict):
        tonic_bolahenk_freq = tonic_symbol = None
        for sym, val in iteritems(tonic_dict):
            if symbol_in in val['makams']:
                tonic_symbol = sym
                tonic_bolahenk_freq = val['bolahenk_freq']
                if not tonic_bolahenk_freq:  # tonic symbol is not known
                    raise KeyError("The tonic of this makam is not known.")
                break
        return tonic_symbol, tonic_bolahenk_freq

    @staticmethod
    def _get_dict(dict_type):
        dict_file = os.path.join(os.path.dirname(
            os.path.abspath(__file__)), 'data', dict_type + '.json')
        return json.load(open(dict_file, 'r'))

    @staticmethod
    def _hz_to_cent(hz_track, ref_freq):
        """--------------------------------------------------------------------
        Converts an array of Hertz values into cents.
        -----------------------------------------------------------------------
        hz_track : The 1-D array of Hertz values
        ref_freq    : Reference frequency for cent conversion
        --------------------------------------------------------------------"""
        hz_track = np.array(hz_track)

        # The 0 Hz values are removed, not only because they are meaningless,
        # but also logarithm of 0 is problematic.
        return np.log2(hz_track[hz_track > 0] / ref_freq) * 1200.0
